src/stash/src/sqlite.rs

```diff
@@ -90,159 +74,43 @@ impl Sqlite {
-             RETURNING key, value, diff",
+              WHERE collection_id = $collection_id AND key = $key",
         )?;
-        let mut insert_stmt = tx.prepare(
-            "INSERT INTO data (collection_id, key, value, time, diff)
-             VALUES ($collection_id, $key, $value, $time, $diff)",
-        )?;
-        let mut drop_stmt = tx.prepare("DELETE FROM data WHERE collection_id = $collection_id")?;
-
-        for collection_id in collections {
-            let since = Self::since_tx(&tx, collection_id)?.into_option();
-            match since {
-                Some(since) => {
-                    let mut updates = consolidation_stmt
-                        .query_and_then(
-                            named_params! {
-                                "$collection_id": collection_id,
-                                "$since": since,
-                            },
-                            |row| {
-                                let key = row.get("key")?;
-                                let value = row.get("value")?;
-                                let diff = row.get("diff")?;
-                                Ok::<_, StashError>(((key, value), since, diff))
-                            },
-                        )?
-                        .collect::<Result<Vec<((Vec<u8>, Vec<u8>), i64, i64)>, _>>()?;
-                    differential_dataflow::consolidation::consolidate_updates(&mut updates);
-                    for ((key, value), time, diff) in updates {
-                        insert_stmt.execute(named_params! {
-                            "$collection_id": collection_id,
-                            "$key": key,
-                            "$value": value,
-                            "$time": time,
-                            "$diff": diff,
-                        })?;
+        for ((key, value), diff) in entries {
+            match diff {
+                Diff::Insert => {
+                    let count = insert_stmt.execute(named_params! {
+                        "$collection_id": collection_id,
+                        "$key": key,
+                        "$value": value,
+                    })?;
+                    if count == 0 {
+                        return Err(StashError::from(InternalStashError::InsertDuplicateKey));
                     }
                 }
-                None => {
-                    drop_stmt.execute(named_params! {
+                Diff::Delete => {
+                    let count = delete_stmt.execute(named_params! {
                         "$collection_id": collection_id,
+                        "$key": key,
                     })?;
+                    if count == 0 {
+                        return Err(StashError::from(InternalStashError::DeleteNotFound));
+                    }
                 }
-            }
+            };
         }
-        drop((consolidation_stmt, insert_stmt, drop_stmt));
+        drop(insert_stmt);
```

Comment:
Why is this `drop` needed right before the functions returns? Won't it be automatically dropped after the return? Also if it includes `insert_stmt` it should probably also include `delete_stmt`.